fvm/src/syscalls/mod.rs

```diff
@@ -74,21 +80,74 @@ pub fn charge_for_exec(
         .map_err(Abort::Fatal)?;
 
     // Determine milligas used, and update the gas tracker.
-    let milligas_used = {
+    let mut exec_gas = {
         let data = ctx.data_mut();
         let last_milligas = mem::replace(&mut data.last_milligas_available, milligas_available);
         // This should never be negative, but we might as well check.
-        last_milligas.saturating_sub(milligas_available)
+        Gas::from_milligas(last_milligas.saturating_sub(milligas_available))
     };
 
-    ctx.data_mut()
-        .kernel
-        .charge_gas("wasm_exec", Gas::from_milligas(milligas_used))
+    let data = ctx.data_mut();
+
+    // Separate the amount of gas charged for memory; this is only makes a difference in tracing.
+    let memory_bytes = data.kernel.limiter_mut().curr_exec_memory_bytes();
+    let memory_delta_bytes = (memory_bytes - data.last_memory_bytes).max(0);
+    let memory_gas = data.kernel.price_list().grow_memory_gas(memory_delta_bytes);
+
+    exec_gas = (exec_gas - memory_gas).max(Gas::zero());
+
+    data.kernel
+        .charge_gas("wasm_exec", exec_gas)
         .map_err(Abort::from_error_as_fatal)?;
 
+    if !memory_gas.is_zero() {
+        data.kernel
+            .charge_gas("wasm_memory_grow", memory_gas)
+            .map_err(Abort::from_error_as_fatal)?;
+    }
+    data.last_memory_bytes = memory_bytes;
+
     Ok(())
 }
 
+/// Charge for the initial memory before a Wasm module is instantiated.
+///
+/// The Wasm instrumentation machinery via [fvm_wasm_instrument::gas_metering::MemoryGrowCost]
+/// only charges for growing the memory _beyond_ the initial amount. It's up to us to make sure
+/// the minimum memory is properly charged for.
+pub fn charge_for_init<K: Kernel>(
+    ctx: &mut impl AsContextMut<Data = InvocationData<K>>,
+    module: &Module,
+) -> crate::kernel::Result<()> {
+    let min_memory_bytes = min_memory_bytes(module)?;
+    let mut ctx = ctx.as_context_mut();
+    let kernel = &mut ctx.data_mut().kernel;
+    let memory_gas = kernel.price_list().init_memory_gas(min_memory_bytes);
+
+    if !memory_gas.is_zero() {
+        kernel.charge_gas("wasm_memory_init", memory_gas)?;
+    }
+
+    // Adjust `last_memory_bytes` so that we don't charge for it again in `charge_for_exec`.
+    ctx.data_mut().last_memory_bytes += min_memory_bytes;
+
+    Ok(())
+}
+
+/// Get the minimum amount of memory required by a module.
+fn min_memory_bytes(module: &Module) -> crate::kernel::Result<usize> {
+    // NOTE: Inside wasmtime this happens slightly differently, by iterating the memory plans:
```

Comment:
Should be covered by validating that the wasm module only exports a single "memory".

Reply:
Currently we only allow a single memory declared [because that's the default](https://docs.rs/wasmtime/latest/wasmtime/struct.InstanceLimits.html#structfield.memories) in `InstanceLimits` which we [don't override](https://github.com/filecoin-project/ref-fvm/pull/1067/files#diff-f6ca4a1c384bf9a1f911827c3974f3f2b25357d937ce1d529732c09404ed6c92R98). As long as the default doesn't change we're fine. Wasmtime [validates](https://github.com/bytecodealliance/wasmtime/blob/v2.0.1/crates/runtime/src/instance/allocator/pooling.rs#L552) that there is no more declared.